components/fxa-client/src/state_machine/internal_machines/mod.rs

```diff
@@ -0,0 +1,211 @@
+                let fxa_error = convert_log_report_error(e);
+                if is_ensure_capabilities && matches!(fxa_error, FxaError::Authentication) {
+                    Ok(Event::EnsureCapabilitiesAuthError)
+                } else {
+                    Ok(Event::CallError)
+                }
+            }
+        })
+    }
+
+    fn make_call_inner(
+        &self,
+        account: &mut FirefoxAccount,
+        device_config: &DeviceConfig,
+    ) -> Result<Event> {
+        Ok(match self {
+            State::GetAuthState => Event::GetAuthStateSuccess {
+                auth_state: account.get_auth_state(),
+            },
+            State::EnsureDeviceCapabilities => {
+                account.ensure_capabilities(&device_config.capabilities)?;
+                Event::EnsureDeviceCapabilitiesSuccess
+            }
+            State::BeginOAuthFlow { scopes, entrypoint } => {
+                let scopes: Vec<&str> = scopes.iter().map(String::as_str).collect();
+                let oauth_url = account.begin_oauth_flow(&scopes, entrypoint)?;
+                Event::BeginOAuthFlowSuccess { oauth_url }
+            }
+            State::BeginPairingFlow {
+                pairing_url,
+                scopes,
+                entrypoint,
+            } => {
+                let scopes: Vec<&str> = scopes.iter().map(String::as_str).collect();
+                let oauth_url = account.begin_pairing_flow(pairing_url, &scopes, entrypoint)?;
+                Event::BeginPairingFlowSuccess { oauth_url }
+            }
+            State::CompleteOAuthFlow { code, state } => {
+                account.complete_oauth_flow(code, state)?;
+                Event::CompleteOAuthFlowSuccess
+            }
+            State::InitializeDevice => {
+                account.initialize_device(
+                    &device_config.name,
+                    device_config.device_type,
+                    &device_config.capabilities,
+                )?;
+                Event::InitializeDeviceSuccess
+            }
+            State::CheckAuthorizationStatus => {
+                let active = account.check_authorization_status()?.active;
+                Event::CheckAuthorizationStatusSuccess { active }
+            }
+            State::Disconnect => {
+                account.disconnect();
+                Event::DisconnectSuccess
+            }
+            state => {
+                return Err(Error::StateMachineLogicError(format!(
+                    "process_call: Don't know how to handle {state}"
+                )))
+            }
+        })
+    }
+}
+
+fn invalid_transition(state: State, event: Event) -> Result<State> {
+    Err(Error::InvalidStateTransition(format!("{state} -> {event}")))
+}
+
+#[cfg(test)]
+struct StateMachineTester {
+    state_machine: Box<dyn InternalStateMachine>,
+    state: State,
+}
+
+#[cfg(test)]
+impl StateMachineTester {
+    fn new<T: InternalStateMachine + 'static>(state_machine: T, event: FxaEvent) -> Self {
```

Comment:
nit: If we're using a generic here, can we make the `StateMachineTester` generic as well and throw away the box?

Or alternatively just make this function take the trait object directly